src/main/java/net/pms/dlna/protocolinfo/ProtocolInfoAttributeName.java

```diff
@@ -0,0 +1,256 @@
+		 * get the singleton instance.
+		 */
+		protected ProtocolInfoAttributeNameFactory() {
+		}
+
+		/**
+		 * Retrieves a {@link ProtocolInfoAttributeName} instance representing
+		 * the attribute name {@code name} from the predefined instances or the
+		 * cache. If no such {@link ProtocolInfoAttributeName} instance exists,
+		 * {@code null} is returned.
+		 *
+		 * @param name the attribute name.
+		 * @return The {@link ProtocolInfoAttributeName} instance or
+		 *         {@code null}.
+		 */
+		public ProtocolInfoAttributeName getAttributeName(String name) {
+
+			// Check for static instances
+			if (isBlank(name)) {
+				return NONE;
+			}
+			name = name.trim().toUpperCase(Locale.ROOT);
+
+			if ("*".equals(name)) {
+				return WILDCARD;
+			}
+
+			// Check for known instances
+			for (KnownProtocolInfoAttributeName knownAttribute : KnownProtocolInfoAttributeName.values()) {
+				if (name.equals(knownAttribute.getName())) {
+					return knownAttribute;
+				}
+			}
+
+			// Check for cached instances
+			instanceCacheLock.readLock().lock();
+			try {
+				for (ProtocolInfoAttributeName cachedAttribute : instanceCache) {
+					if (name.equals(cachedAttribute.getName())) {
+						return cachedAttribute;
+					}
+				}
+			} finally {
+				instanceCacheLock.readLock().unlock();
+			}
+			return null;
+		}
+
+		/**
+		 * Creates or retrieves a {@link ProtocolInfoAttributeName} instance
+		 * representing the attribute name {@code name}. Any existing predefined
+		 * or cached instances are first checked, and if none are found for
+		 * {@code value} a new instance is created.
+		 *
+		 * @param name the profile name.
+		 * @return The {@link ProtocolInfoAttributeName} instance.
+		 */
+		public ProtocolInfoAttributeName createAttributeName(String name) {
+
+			ProtocolInfoAttributeName instance = getAttributeName(name);
+			if (instance != null) {
+				return instance;
+			}
+
+			// Null values will already have been handled above
+			name = name.trim().toUpperCase(Locale.ROOT);
+
+			// Prepare to create a new instance
+			instanceCacheLock.writeLock().lock();
+			try {
+				// Check cache again as it could have been added while the lock
+				// was released
+				for (ProtocolInfoAttributeName cachedAttribute : instanceCache) {
+					if (name.equals(cachedAttribute.getName())) {
+						return cachedAttribute;
+					}
+				}
+
+				// None was found, create the instance
+				instance = new GenericProtocolInfoAttributeName(name);
+				instanceCache.add(instance);
+				LOGGER.trace("ProtocolInfoAttributeNameFactory added unknown attribute name \"{}\"", name);
+				return instance;
+			} finally {
+				instanceCacheLock.writeLock().unlock();
+			}
+		}
+	}
+
+	/**
+	 * This contains predefined, known {@code protocolInfo} attribute names.
+	 */
+	public enum KnownProtocolInfoAttributeName implements ProtocolInfoAttributeName {
+		// The order is important for protocolInfo
+
```

Comment:
Really ? i didn't know. Where did you get that information from ? at least the section reference if not the article.

Good job.